File: clickhouse-ext/clickhousephp.c
#include <php.h>
#include <Zend/zend_API.h>
#include <Zend/zend_hash.h>
#include <Zend/zend_types.h>
#include <Zend/zend_exceptions.h>
#include <ext/spl/spl_exceptions.h>
#include <stddef.h>
#include <string.h>

#include "clickhousephp.h"
#include "clickhousephp_arginfo.h"
#include "_cgo_export.h"

// Helper: if result starts with "ERROR: ", throw RuntimeException and return 1.
static int ch_throw_on_error(zend_string *result) {
    if (result && ZSTR_LEN(result) > 7 && memcmp(ZSTR_VAL(result), "ERROR: ", 7) == 0) {
        zend_throw_exception(spl_ce_RuntimeException, ZSTR_VAL(result) + 7, 0);
        zend_string_release(result);
        return 1;
    }
    return 0;
}

PHP_MINIT_FUNCTION(clickhousephp) {
    return SUCCESS;
}

zend_module_entry clickhousephp_module_entry = {STANDARD_MODULE_HEADER,
                                         "clickhousephp",
                                         ext_functions,
                                         PHP_MINIT(clickhousephp),
                                         NULL, NULL, NULL, NULL,
                                         "1.0.0",
                                         STANDARD_MODULE_PROPERTIES};

PHP_FUNCTION(clickhouse_connect)
{
    zend_string *dsn = NULL;
    ZEND_PARSE_PARAMETERS_START(1, 1)
        Z_PARAM_STR(dsn)
    ZEND_PARSE_PARAMETERS_END();
    zend_string *result = clickhouse_connect(dsn);
    if (ch_throw_on_error(result)) { RETURN_THROWS(); }
    if (result) { RETURN_STR(result); }
    RETURN_EMPTY_STRING();
}

PHP_FUNCTION(clickhouse_disconnect)
{
    if (zend_parse_parameters_none() == FAILURE) { RETURN_THROWS(); }
    zend_string *result = clickhouse_disconnect();
    if (ch_throw_on_error(result)) { RETURN_THROWS(); }
    if (result) { RETURN_STR(result); }
    RETURN_EMPTY_STRING();
}

PHP_FUNCTION(clickhouse_insert)
{
    zend_string *table = NULL;
    zval *values = NULL;
    zval *columns = NULL;
    zval *options = NULL;
    ZEND_PARSE_PARAMETERS_START(2, 4)
        Z_PARAM_STR(table)
        Z_PARAM_ARRAY(values)
        Z_PARAM_OPTIONAL
        Z_PARAM_ARRAY_OR_NULL(columns)
        Z_PARAM_ARRAY_OR_NULL(options)
    ZEND_PARSE_PARAMETERS_END();
    zend_string *result = clickhouse_insert(table, values, columns, options);
    if (ch_throw_on_error(result)) { RETURN_THROWS(); }
    // Also catch "Insert error" and "Send error" prefixes from Go
    if (result && (
        (ZSTR_LEN(result) > 13 && memcmp(ZSTR_VAL(result), "Insert error", 12) == 0) ||
        (ZSTR_LEN(result) > 10 && memcmp(ZSTR_VAL(result), "Send error", 10) == 0)
    )) {
        zend_throw_exception(spl_ce_RuntimeException, ZSTR_VAL(result), 0);
        zend_string_release(result);
        RETURN_THROWS();
    }
    if (result) { RETURN_STR(result); }
    RETURN_EMPTY_STRING();
}

PHP_FUNCTION(clickhouse_exec)
{
    zend_string *query = NULL;
    zval *params = NULL;
    zval *options = NULL;
    ZEND_PARSE_PARAMETERS_START(1, 3)
        Z_PARAM_STR(query)
        Z_PARAM_OPTIONAL
        Z_PARAM_ARRAY_OR_NULL(params)
        Z_PARAM_ARRAY_OR_NULL(options)
    ZEND_PARSE_PARAMETERS_END();
    zend_string *result = clickhouse_exec(query, params, options);
    if (ch_throw_on_error(result)) { RETURN_THROWS(); }
    if (result) { RETURN_STR(result); }
    RETURN_EMPTY_STRING();
}

PHP_FUNCTION(clickhouse_query_array)
{
    zend_string *query = NULL;
    zval *params = NULL;
    zval *options = NULL;
    ZEND_PARSE_PARAMETERS_START(1, 3)
        Z_PARAM_STR(query)
        Z_PARAM_OPTIONAL
        Z_PARAM_ARRAY_OR_NULL(params)
        Z_PARAM_ARRAY_OR_NULL(options)
    ZEND_PARSE_PARAMETERS_END();
    zend_array *result = clickhouse_query_array(query, params, options);
    if (result == NULL) {
        zend_string *err = clickhouse_get_last_error();
        const char *msg = err ? ZSTR_VAL(err) : "ClickHouse query failed";
        zend_throw_exception(spl_ce_RuntimeException, msg, 0);
        if (err) { zend_string_release(err); }
        RETURN_THROWS();
    }
    RETURN_ARR(result);
}

PHP_FUNCTION(clickhouse_query_cursor)
{
    zend_string *query = NULL;
    zval *params = NULL;
    zval *options = NULL;
    ZEND_PARSE_PARAMETERS_START(1, 3)
        Z_PARAM_STR(query)
        Z_PARAM_OPTIONAL
        Z_PARAM_ARRAY_OR_NULL(params)
        Z_PARAM_ARRAY_OR_NULL(options)
    ZEND_PARSE_PARAMETERS_END();
    int64_t id = clickhouse_query_cursor(query, params, options);
    if (id < 0) {
        zend_string *err = clickhouse_get_last_error();
        const char *msg = err ? ZSTR_VAL(err) : "ClickHouse cursor open failed";
        zend_throw_exception(spl_ce_RuntimeException, msg, 0);
        if (err) { zend_string_release(err); }
        RETURN_THROWS();
    }
    RETURN_LONG((zend_long)id);
}

PHP_FUNCTION(clickhouse_cursor_fetch)
{
    zend_long id = 0;
    zend_long max_rows = 10000;
    ZEND_PARSE_PARAMETERS_START(1, 2)
        Z_PARAM_LONG(id)
        Z_PARAM_OPTIONAL
        Z_PARAM_LONG(max_rows)
    ZEND_PARSE_PARAMETERS_END();
    zend_array *result = clickhouse_cursor_fetch((int64_t)id, (int64_t)max_rows);
    if (result == NULL) {
        zend_string *err = clickhouse_get_last_error();
        const char *msg = err ? ZSTR_VAL(err) : "ClickHouse cursor fetch failed";
        zend_throw_exception(spl_ce_RuntimeException, msg, 0);
        if (err) { zend_string_release(err); }
        RETURN_THROWS();
    }
    RETURN_ARR(result);
}

PHP_FUNCTION(clickhouse_cursor_close)
{
    zend_long id = 0;
    ZEND_PARSE_PARAMETERS_START(1, 1)
        Z_PARAM_LONG(id)
    ZEND_PARSE_PARAMETERS_END();
    zend_string *result = clickhouse_cursor_close((int64_t)id);
    if (ch_throw_on_error(result)) { RETURN_THROWS(); }
    if (result) { RETURN_STR(result); }
    RETURN_EMPTY_STRING();
}

PHP_FUNCTION(clickhouse_ping)
{
    if (zend_parse_parameters_none() == FAILURE) { RETURN_THROWS(); }
    zend_string *result = clickhouse_ping();
    if (ch_throw_on_error(result)) { RETURN_THROWS(); }
    if (result) { RETURN_STR(result); }
    RETURN_EMPTY_STRING();
}

PHP_FUNCTION(clickhouse_server_version)
{
    if (zend_parse_parameters_none() == FAILURE) { RETURN_THROWS(); }
    zend_string *result = clickhouse_server_version();
    if (ch_throw_on_error(result)) { RETURN_THROWS(); }
    if (result) { RETURN_STR(result); }
    RETURN_EMPTY_STRING();
}

PHP_FUNCTION(clickhouse_batch_begin)
{
    zend_string *table = NULL;
    zval *columns = NULL;
    zval *options = NULL;
    ZEND_PARSE_PARAMETERS_START(1, 3)
        Z_PARAM_STR(table)
        Z_PARAM_OPTIONAL
        Z_PARAM_ARRAY_OR_NULL(columns)
        Z_PARAM_ARRAY_OR_NULL(options)
    ZEND_PARSE_PARAMETERS_END();
    int64_t id = clickhouse_batch_begin(table, columns, options);
    if (id < 0) {
        zend_string *err = clickhouse_get_last_error();
        const char *msg = err ? ZSTR_VAL(err) : "ClickHouse batch open failed";
        zend_throw_exception(spl_ce_RuntimeException, msg, 0);
        if (err) { zend_string_release(err); }
        RETURN_THROWS();
    }
    RETURN_LONG((zend_long)id);
}

PHP_FUNCTION(clickhouse_batch_append)
{
    zend_long id = 0;
    zval *values = NULL;
    ZEND_PARSE_PARAMETERS_START(2, 2)
        Z_PARAM_LONG(id)
        Z_PARAM_ARRAY(values)
    ZEND_PARSE_PARAMETERS_END();
    zend_string *result = clickhouse_batch_append((int64_t)id, values);
    if (ch_throw_on_error(result)) { RETURN_THROWS(); }
    if (result) { RETURN_STR(result); }
    RETURN_EMPTY_STRING();
}

PHP_FUNCTION(clickhouse_batch_flush)
{
    zend_long id = 0;
    ZEND_PARSE_PARAMETERS_START(1, 1)
        Z_PARAM_LONG(id)
    ZEND_PARSE_PARAMETERS_END();
    zend_string *result = clickhouse_batch_flush((int64_t)id);
    if (ch_throw_on_error(result)) { RETURN_THROWS(); }
    if (result) { RETURN_STR(result); }
    RETURN_EMPTY_STRING();
}

PHP_FUNCTION(clickhouse_batch_send)
{
    zend_long id = 0;
    ZEND_PARSE_PARAMETERS_START(1, 1)
        Z_PARAM_LONG(id)
    ZEND_PARSE_PARAMETERS_END();
    zend_string *result = clickhouse_batch_send((int64_t)id);
    if (ch_throw_on_error(result)) { RETURN_THROWS(); }
    if (result) { RETURN_STR(result); }
    RETURN_EMPTY_STRING();
}

PHP_FUNCTION(clickhouse_batch_abort)
{
    zend_long id = 0;
    ZEND_PARSE_PARAMETERS_START(1, 1)
        Z_PARAM_LONG(id)
    ZEND_PARSE_PARAMETERS_END();
    zend_string *result = clickhouse_batch_abort((int64_t)id);
    if (ch_throw_on_error(result)) { RETURN_THROWS(); }
    if (result) { RETURN_STR(result); }
    RETURN_EMPTY_STRING();
}

PHP_FUNCTION(clickhouse_async_insert)
{
    zend_string *query = NULL;
    bool wait = 1;
    zval *params = NULL;
    zval *options = NULL;
    ZEND_PARSE_PARAMETERS_START(1, 4)
        Z_PARAM_STR(query)
        Z_PARAM_OPTIONAL
        Z_PARAM_BOOL(wait)
        Z_PARAM_ARRAY_OR_NULL(params)
        Z_PARAM_ARRAY_OR_NULL(options)
    ZEND_PARSE_PARAMETERS_END();
    zend_string *result = clickhouse_async_insert(query, wait ? 1 : 0, params, options);
    if (ch_throw_on_error(result)) { RETURN_THROWS(); }
    if (result) { RETURN_STR(result); }
    RETURN_EMPTY_STRING();
}

File: stubs/clickhousephp.php
<?php

/**
 * IDE stubs for the `clickhousephp` FrankenPHP extension.
 *
 * These declarations exist only so editors and static analysers (PhpStorm,
 * Intelephense, PHPStan, Psalm…) can resolve the extension's global functions
 * with full type information. The file is NOT autoloaded — the real functions
 * are provided by the compiled FrankenPHP binary at runtime, so requiring this
 * file would trigger a redeclaration error. Install it with:
 *
 *     composer require --dev jhondermarck/frankenphp-clickhouse-stubs
 *
 * Every function throws \RuntimeException on error; for server-side errors the
 * exception code carries the ClickHouse error code (see getCode()).
 */

/**
 * Open (or replace) the default, process-wide connection pool.
 *
 * @param string $dsn clickhouse://[user[:pass]@]host:9000[,host2…]/db[?param=value]
 * @return string "Ok"
 * @throws \RuntimeException on a bad DSN, unreachable host, or auth failure
 */
function clickhouse_connect(string $dsn): string {}

/**
 * Close the default connection pool.
 *
 * @return string "Ok"
 * @throws \RuntimeException when no default connection is open
 */
function clickhouse_disconnect(): string {}

/**
 * Open an extra named connection and return its handle, for routing calls to a
 * second cluster/database via the `connection` option.
 *
 * @param string $dsn see {@see clickhouse_connect()}
 * @return int a connection handle (> 0)
 * @throws \RuntimeException on connection failure
 */
function clickhouse_open(string $dsn): int {}

/**
 * Close a connection opened with {@see clickhouse_open()}.
 *
 * @param int $connection handle from clickhouse_open()
 * @return string "Ok"
 * @throws \RuntimeException if the handle is unknown
 */
function clickhouse_close(int $connection): string {}

/**
 * Run a SELECT and return every row as a native PHP array (no JSON).
 *
 * @param string $query SQL with optional {name:Type} placeholders
 * @param array<string,mixed>|list<mixed>|null $params named (assoc) or positional (list) bindings
 * @param array{connection?:int,settings?:array<string,mixed>,query_id?:string,timeout?:string}|null $options
 * @return list<array<string,mixed>> rows, each an associative array keyed by column
 * @throws \RuntimeException on query error or an unsupported column type
 */
function clickhouse_query_array(string $query, ?array $params = null, ?array $options = null): array {}

/**
 * Open a streaming cursor for a SELECT — bounded memory, fetched in chunks.
 * Always release it with {@see clickhouse_cursor_close()}.
 *
 * @param string $query SQL with optional {name:Type} placeholders
 * @param array<string,mixed>|list<mixed>|null $params bindings
 * @param array{connection?:int,settings?:array<string,mixed>,query_id?:string,timeout?:string}|null $options
 * @return int a cursor handle (> 0)
 * @throws \RuntimeException on query error
 */
function clickhouse_query_cursor(string $query, ?array $params = null, ?array $options = null): int {}

/**
 * Fetch the next chunk of rows from a cursor. Returns an empty array once the
 * result is exhausted.
 *
 * @param int $cursor handle from clickhouse_query_cursor()
 * @param int $max_rows maximum rows to return this call
 * @return list<array<string,mixed>> rows (empty when exhausted)
 * @throws \RuntimeException on a mid-stream failure or an unknown cursor
 */
function clickhouse_cursor_fetch(int $cursor, int $max_rows = 10000): array {}

/**
 * Close a cursor and release its pooled connection.
 *
 * @param int $cursor handle from clickhouse_query_cursor()
 * @return string "Ok"
 */
function clickhouse_cursor_close(int $cursor): string {}

/**
 * Execute a statement that returns no rows (DDL, TRUNCATE, KILL QUERY…).
 *
 * @param string $query SQL with optional {name:Type} placeholders
 * @param array<string,mixed>|list<mixed>|null $params bindings
 * @param array{connection?:int,settings?:array<string,mixed>,query_id?:string,timeout?:string}|null $options
 * @return string "Ok"
 * @throws \RuntimeException on error
 */
function clickhouse_exec(string $query, ?array $params = null, ?array $options = null): string {}

/**
 * Batch INSERT from a fully-materialized payload. Accepts flat values (with
 * $columns), nested rows, or associative rows (columns inferred from keys).
 *
 * @param string $table target table (validated identifier, never user input)
 * @param array $values flat list, list of rows, or list of assoc rows
 * @param list<string>|null $columns column names (required for flat/assoc input)
 * @param array{connection?:int,settings?:array<string,mixed>,query_id?:string,timeout?:string}|null $options
 * @return string "Ok"
 * @throws \RuntimeException on a value/column mismatch or server error
 */
function clickhouse_insert(string $table, array $values, ?array $columns = null, ?array $options = null): string {}

/**
 * Begin an incremental insert batch for unbounded-size writes; returns a
 * handle. Append chunks, flush periodically, then send or abort. The pooled
 * connection is held from begin to send/abort.
 *
 * @param string $table target table
 * @param list<string>|null $columns column names (required for flat/assoc appends)
 * @param array{connection?:int,settings?:array<string,mixed>,query_id?:string,timeout?:string}|null $options
 * @return int a batch handle (> 0)
 * @throws \RuntimeException on error
 */
function clickhouse_batch_begin(string $table, ?array $columns = null, ?array $options = null): int {}

/**
 * Append rows to an open batch (buffered client-side).
 *
 * @param int $batch handle from clickhouse_batch_begin()
 * @param array $values same shapes as {@see clickhouse_insert()} $values
 * @return string "Ok"
 * @throws \RuntimeException if the batch is closed or the rows mismatch
 */
function clickhouse_batch_append(int $batch, array $values): string {}

/**
 * Ship buffered rows to the server, keeping the batch open (memory stays
 * bounded on both sides).
 *
 * @param int $batch handle from clickhouse_batch_begin()
 * @return string "Ok"
 * @throws \RuntimeException on flush failure (the batch is then closed)
 */
function clickhouse_batch_flush(int $batch): string {}

/**
 * Finalize a batch: flush any buffered rows and commit, then close it.
 *
 * @param int $batch handle from clickhouse_batch_begin()
 * @return string "Ok"
 * @throws \RuntimeException on send failure
 */
function clickhouse_batch_send(int $batch): string {}

/**
 * Discard an open batch. Rows already flushed remain on the server; buffered
 * rows are dropped.
 *
 * @param int $batch handle from clickhouse_batch_begin()
 * @return string "Ok"
 */
function clickhouse_batch_abort(int $batch): string {}

/**
 * Server-side buffered INSERT for high-frequency small writes.
 *
 * @param string $query INSERT … VALUES (…) with optional {name:Type} placeholders
 * @param bool $wait true = return once the server flushed the buffer (durable);
 *                   false = fire-and-forget
 * @param array<string,mixed>|list<mixed>|null $params bindings
 * @param array{connection?:int,settings?:array<string,mixed>,query_id?:string,timeout?:string}|null $options
 * @return string "Ok"
 * @throws \RuntimeException on error
 */
function clickhouse_async_insert(string $query, bool $wait = true, ?array $params = null, ?array $options = null): string {}

/**
 * Ping a connection.
 *
 * @param int|null $connection handle from clickhouse_open() (null = default)
 * @return string "Ok"
 * @throws \RuntimeException if the server is unreachable
 */
function clickhouse_ping(?int $connection = null): string {}

/**
 * Return the server version as "major.minor.patch".
 *
 * @param int|null $connection handle from clickhouse_open() (null = default)
 * @return string e.g. "26.5.5"
 * @throws \RuntimeException on error
 */
function clickhouse_server_version(?int $connection = null): string {}

/**
 * Return a process-wide runtime snapshot (no server round-trip) for health
 * checks and leak diagnosis.
 *
 * @return array{
 *   connected:int,
 *   uptime_seconds:int,
 *   server_version:string,
 *   named_connections:int,
 *   handles:array{cursors_open:int,batches_open:int,idle_ttl_seconds:int,last_reap_unix:int,last_reap_count:int},
 *   pool:array{open?:int,idle?:int,max_open_conns?:int,max_idle_conns?:int},
 *   counters:array{queries:int,inserts:int,execs:int,async_inserts:int,cursors_opened:int,batches_opened:int,errors:int}
 * }
 */
function clickhouse_stats(): array {}
